### ISSUE TITLE
why is EUI-64 encoded as DNS-ID

### ISSUE THREAD

**Issue by mcr** [thomas-fossati/draft-tls13-iot#65]
````
If the EUI-64 format is used to identify the subject of an end entity
certificate, it MUST be encoded in a subjectAltName of type DNS-ID as a string
of the form `HH-HH-HH-HH-HH-HH-HH-HH` where 'H' is one of the symbols '0'-'9'
or 'A'-'F'.
````

this seems really odd.  I was looking for a SN, or SAN that would encode EUI64, and I feel surprised not to find one.

**Comment by thomas-fossati** [comment]
This seems to come from [RFC7925 §4.4.2](https://datatracker.ietf.org/doc/html/rfc7925#section-4.4.2) and the requirement for DNS-compatibility is in order to support SNI for server EE certs.

**Comment by mcr** [comment]
> @thomas-fossati that reference to RFC 7925 says to put it in the CN, _not_ the SAN. This is a really bad idea. It's not a hostname, do not try to make it look like one. Define a URI scheme and use that in the SAN entry.

fair enough suggestion as to where to put it.  URI scheme or otherName, I don't care that much.
But neither fits into SNI.

**Comment by richsalz** [comment]
Technically RFC 6066 has an escape hatch, this sentence near the top of page 7: "TLS MAY treat provided server names as opaque data and pass the names and types to the application." Although the following sentences say that it should be a FQDN, they don't use 2119 MUST, etc., language.

**Comment by boaks** [comment]
I'm little mixed up.

There are two points mentions the SAN in x509  and the SNI, both are related, but my feeling is, that we should be clear on the usage.

A device certificate will mainly be used by a DTLS client to authenticate itself, right?
Or do we assume a DTLS role exchange, where the device acts as DTLS server?

If the device acts as DTLS client, then I hardly see the relation to SNI (as that name already implies, this is about the server's name).

So, before continue, please clarify, if it's considered, that a device acts as DTLS server.

**Comment by boaks** [comment]
>  I was looking for a SN, or SAN that would encode EUI64, and I feel surprised not to find one.

I may be wrong, but RFC 7925 is intended to show to use what is "already specified" rather then introduce "new specifications".
If SAN doesn't support EUI64, then that's a issue of that, but not of RFC 7925.

**Comment by thomas-fossati** [comment]
> @thomas-fossati that reference to RFC 7925 says to put it in the CN, _not_ the SAN.

Rich, I'm not sure I'm getting your point.  The referenced text says to put it in the CN _or_ the SAN: _"For client certificates, the identifier used in the **SubjectAltName** or in the **leftmost CN component** of subject name MUST be an EUI-64."_

> This is a really bad idea. It's not a hostname, do not try to make it look like one. Define a URI scheme and use that in the SAN entry.

Yes, it looks bad.  However, (likely due to a lack of imagination) it's the only solution we could devise to work around the SNI limitation.

**Comment by boaks** [comment]
FMPOV, there are different assumption applied to client- and server-certificates.
I guess, the intention of this convention is to use it mainly for client-certificates.
If used for server-certificates, then this will only be the case in iot-specific use-cases.

**Comment by thomas-fossati** [comment]
> > Yes, this is scoped to EE certs issued to IoT (D)TLS servers.
> 
> If so, will RFC 7925
> 
> "Note that the SNI extension cannot be used in this case since SNI does not offer the ability to convey a 64-bit Extended Unique Identifier (EUI-64)."
> 
> still apply for "draft-tls13-iot"?

It looks like this would be one of the "updates" to 7925 -- depending, of course, on how this conversation ends :-)

**Comment by boaks** [comment]
OK, I understand.

So it seems to be the question, 
if (even if I don't know, who will work on that) EUI64 should be supported directly in x509 and SNI is then still not to be used. 
Or if the approach is to "encode" the EUI64 in a way, that SNI has a chance.

My 2 cents:
The D2D use-case is pretty rare ... maybe it gets more used in the future 
(but at least the last request for D2D, which reached me last week, then turned out to use a "broker").
I'm not sure, if such a setup will use x509 anyway.
Therefore I guess, it will not get the attention required to add EUI64 to x509.
And also therefore, I'm not sure, if using then SNI pays-off, at least in the sense of "interoperable out-of-the-box" ;-).

**Comment by thomas-fossati** [comment]
@boaks thanks a lot for your input.

**Comment by richsalz** [comment]
If you put it in the CN component, I have no concerns. If you say "put it in the SAN field as a dNSName" I am bothered by that. I hope you make it clear that the CN is the only place to put it.  (In fact, that makes the 9525 case stronger, which says that the CN field has no typing associated with it :)

I do not care about the SNI value in the TLS handshake.

**Comment by primetomas** [comment]
> **leftmost CN component**

What does letfmost CN component refer to? The Subject DN is an ASN.1 sequence so there is nothing leftmost in a subjectDN. Depending on order of DN components is a bad practice and should be avoided, different tools may print it in different order. If you want to use the CN I would specify that the CN must be a specific value, and there must be only one.

**Comment by mcr** [comment]
@Achim asked about what is going on client vs server.
The answer is subtle, and he really got the entire client authentication
storey.  No, there is no turn around client/server, but rather one can have
interactions which are device to device (one a client, one a server), and
also have device to cloud interactions (the device a client, the cloud a
server).

While I guess SNI has that escape hatch which lets us put non-DNS things into
the SNI.  This lets the server pick the right keypair to respond with.
The problem is then, having picked a certificate, the client has to validate
that it got a reasonable certificate.  9525 says we do DNS-ID checks only on
the SAN, so we wind up putting junk into the SAN.

My answer is: don't use (have server ignore) SNI for device to device.

**Comment by boaks** [comment]
@mcr 

"@boaks" and "@achim" are the same person ;-).

> My answer is: don't use (have server ignore) SNI for device to device.

OK, that answer about SNI I do understand.

But I still don't get your intention about the EUI64 and certificates.

My question from the mailing list:

> So, please:
> Is it about direct EUI64 support in x509?
> Or about omit EUI64 in device certificates?

**Comment by boaks** [comment]
For those who missed the discussion on the mailing-list:

[UTA mailing-list, Michael](https://mailarchive.ietf.org/arch/msg/uta/-N4PkGJGW6sLfZppIuBpQUrseIo/)

```
    > So, please: Is it about direct EUI64 support in x509?  Or about omit
    > EUI64 in device certificates?

This is about what SNI supports vs what X509 supports.
```

[UTA mailing-list, Valery](https://mailarchive.ietf.org/arch/msg/uta/UG1JIHXBCVbYeiXanEQsFh3ee6U/)

```
Hmm, RFC 6066 actually says:

    For backward compatibility, all future data structures associated with
    new NameTypes MUST begin with a 16-bit length field.

In my reading of this requirement, it allows servers to effectively skip unknown Name Types.
Am I missing something?

Regards,
Valery.
```

**Comment by hannestschofenig** [comment]
Closed because of PR#71
